### DISCUSSION TITLE
Unable to pass a vector list as a new parameter

### DISCUSSION THREAD

**Discussion by uttamcadambi07** [mdolab/dafoam#406] [Bug] [answered]
Hi,

DAFoam version: v3.0.3

I have been trying to add a new parameter and followed the instructions from DAFoam workshop 2022 on adding a new parameter.

I would like this parameter in DAObjFunc class so I create a new class DAMyObjFunc.C and *.H and add this line in it

```
List<vector> reference_velocity = daOption_.getOption< List<vector> >("reference_velocity");
```

Compilation is fine. I set in runScript.py for the curved cube case to test if everything is working fine

```
daOptions = {
    "solverName": "DASimpleFoam",
    "primalMinResTol": 1.0e-8,
    "reference_velocity": velocity_reference,
.
.
.
}
```
and other options that are necessary where "velocity_reference" is a list of lists (each element of the main list has three elements in itself). I get this error when I run the case

```
--> FOAM FATAL ERROR: 
Type: <list> for reference_velocity list is not supported! Options are: str, int, bool, and float!

    From function pyDict2OFDict
    in file DAUtility/DAUtility.C at line 184.

```

So does this mean that the way I have declared the variable in the DAMyObjFunc class is wrong or do I need to extend the functionality of the DAUtility class to include vector lists? Please advise.

**Comment by friedenhe** [comment]
Yeah, the errors says `List<vector>` is not supported. Use `List<scalar>` instead

**Reply by uttamcadambi07** [reply]
Okay thank you.

I will explain this for the benefit of others.

I use:
```
scalarList ux_ref = daOption_.getOption<scalarList>("ux_ref");
```
just as you recommended. But there is another issue.

In `$DAFOAM/src/adjoint/DAUtility/DAUtility.C`, there is no implementation for the case when the input data type from Python is of type `<numpy.float64>`. DAFoam only accepts native python types, for example - float, int, bool, str. So there are two ways to tackle this problem.

1 Define the else if block in `../DAUtility.C'  say after Line 176 to tackle non-native Python types. For example: I tried implementing this code
```
 else if (tmpTypeWord == "numpy.float64")
                {
                    double valSet = PyFloat_AsDouble(valueListJ);  // use PyFloat_AsDouble instead of PyFloat_AS_DOUBLE for float64 type
                    valSetScalar[j] = valSet;
                }
```
it was not successful. Maybe someone else can check it out.

2. Cast your Python numpy array into a primitive type. I did this using the code `ux_ref = list(ux_ref.tolist())` where my `ux_ref` is basically a numpy array of type `<numpy.float64>` which I have converted to a native data type. This works for me.

**Comment by friedenhe** [comment, marked as answer]
Great to hear that it is working. Yeah, the best way is to convert to Python's native datatype
